### PR TITLE
Improvements to astbuilder

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
Pulled out of #378, a collection of changes to the `astbuilder` package, including:

* Automatic cloning of passed parameters to avoid errors writing generated code
* Creation of `lists.go` and `maps.go` to group functions for easier discovery
* New methods `Dereference()` `Selector()` and `NotEqual()`
* Helper methods to clone slices of `dst.Expr` and `dst.Stmt`

### CODE REVIEW

File: hack/generator/pkg/astbuilder/builder.go [hunk 1]
```diff
@@ -257,7 +222,15 @@ func FormatError(fmtPackage string, formatString string, args ...dst.Expr) dst.E
 func AddrOf(exp dst.Expr) *dst.UnaryExpr {
 	return &dst.UnaryExpr{
 		Op: token.AND,
-		X:  exp,
+		X:  dst.Clone(exp).(dst.Expr),
+	}
+}
+
+// Dereference returns a statement that dereferences the pointer returned by the provided expression
```
Comment (matthchr) [inline hack/generator/pkg/astbuilder/builder.go:229]: minor: include sample code snippet too?

File: hack/generator/pkg/astbuilder/builder.go [hunk 2]
```diff
@@ -283,3 +256,36 @@ func QualifiedTypeName(pkg string, name string) *dst.SelectorExpr {
 		Sel: dst.NewIdent(name),
 	}
 }
+
+func Selector(expr dst.Expr, name string) *dst.SelectorExpr {
```
Comment (matthchr) [inline hack/generator/pkg/astbuilder/builder.go:260]: missing comments and examples on this and a few methods below

File: hack/generator/pkg/astbuilder/builder.go [hunk 3]
```diff
@@ -283,3 +256,36 @@ func QualifiedTypeName(pkg string, name string) *dst.SelectorExpr {
...
+
+func NotEqual(lhs dst.Expr, rhs dst.Expr) *dst.BinaryExpr {
+	return &dst.BinaryExpr{
+		X:  dst.Clone(lhs).(dst.Expr),
+		Op: token.NEQ,
+		Y:  dst.Clone(rhs).(dst.Expr),
+	}
+}
+
+func cloneExprSlice(exprs []dst.Expr) []dst.Expr {
```
Comment (matthchr) [inline hack/generator/pkg/astbuilder/builder.go:275]: Another opportunity for generics!

File: hack/generator/pkg/astbuilder/lists.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// AppendList returns a statement for a list append, like:
+//     <lhs> = append(<lhs>, <rhs>)
+func AppendList(lhs dst.Expr, rhs dst.Expr) dst.Stmt {
+	return SimpleAssignment(
+		dst.Clone(lhs).(dst.Expr),
+		token.ASSIGN,
+		CallFunc("append", dst.Clone(lhs).(dst.Expr), dst.Clone(rhs).(dst.Expr)))
+}
+
+func IterateOverList(item string, list dst.Expr, statements ...dst.Stmt) *dst.RangeStmt {
```
Comment (matthchr) [inline hack/generator/pkg/astbuilder/lists.go:34]: Missing comments here and below

File: hack/generator/pkg/astbuilder/maps.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	return SimpleAssignment(
+		&dst.IndexExpr{
+			X:     dst.Clone(m).(dst.Expr),
+			Index: dst.Clone(key).(dst.Expr),
+		},
+		token.ASSIGN,
+		dst.Clone(rhs).(dst.Expr))
+}
+
+func IterateOverMapWithValue(key string, item string, list dst.Expr, statements ...dst.Stmt) *dst.RangeStmt {
```
Comment (matthchr) [inline hack/generator/pkg/astbuilder/maps.go:39]: Missing comment

**Review by matthchr — COMMENTED** [review]
Looks good but missing a few comments

File: hack/generator/pkg/astbuilder/builder.go (hunk 1 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astbuilder/builder.go:250]: Done.

File: hack/generator/pkg/astbuilder/builder.go (hunk 2 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astbuilder/builder.go:292]: Added.

File: hack/generator/pkg/astbuilder/lists.go (hunk 4 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astbuilder/lists.go:45]: Added.

File: hack/generator/pkg/astbuilder/maps.go (hunk 5 above)
Comment (theunrepentantgeek) [reply hack/generator/pkg/astbuilder/maps.go]: Added.

**Review by theunrepentantgeek — COMMENTED** [review]
Thanks @matthchr, I think I've addressed all your feedback.

**Review by matthchr — APPROVED** [review]
LGTM!

**Pull request merged**